File: source/tools/rmgen/map.cpp
#include "stdafx.h"
#include "rmgen.h"
#include "map.h"
#include "object.h"
#include "pmp_file.h"

using namespace std;

Map::Map(int size, Terrain* baseTerrain, float baseHeight) {
	if(size<0 || size>1024) {
		JS_ReportError(cx, "init: map size out of range");
	}
	else if(size%16 != 0) {
		JS_ReportError(cx, "init: map size must be divisible by 16");
	}

	this->size = size;

	texture = new int*[size];
	for(int i=0; i<size; i++) {
		texture[i] = new int[size];
	}

	terrainObjects = new vector<Object*>*[size];
	for(int i=0; i<size; i++) {
		terrainObjects[i] = new vector<Object*>[size];
	}

	area = new Area**[size];
	for(int i=0; i<size; i++) {
		area[i] = new Area*[size];
		for(int j=0; j<size; j++) {
			area[i][j] = 0;
		}
	}

	height = new float*[size+1];
	for(int i=0; i<size+1; i++) {
		height[i] = new float[size+1];
		for(int j=0; j<size+1; j++) {
			height[i][j] = baseHeight;
		}
	}

	for(int i=0; i<size; i++) {
		for(int j=0; j<size; j++) {
			baseTerrain->place(this, i, j);
		}
	}
}

Map::Map(string fileName, int loadLevel)
{
	const int LOAD_NOTHING = 0;
	const int LOAD_TERRAIN = 1<<0;
	const int LOAD_INTERACTIVES = 1 << 1;
	const int LOAD_NONINTERACTIVES = 1 << 2;
	//const int SOMETHINGELSE = 1 << 3;


	// HACK, this should probably be in a struct and be shared with the code in output.cpp
	pmp_header hdr;
	u32 map_size;

	//HACK, also in rmgen.cpp
	const string SCENARIO_PATH = "../data/mods/public/maps/scenarios/";

	std::string pmpFile = SCENARIO_PATH + fileName + ".pmp";
	std::string xmlFile = SCENARIO_PATH + fileName + ".xml";

	if (loadLevel & LOAD_TERRAIN)
	{

		FILE* f = fopen(pmpFile.c_str(), "rb");

		fread(&hdr, sizeof(pmp_header), 1, f);

		/*TODO

		if (hdr.marker !== "PSMP")
			JS_ReportError(cx, "");

		if (hdr.version !== 4)
			JS_ReportError(cx, ""); */
		

		fread(&map_size, sizeof(int), 1, f);

		size = map_size * 16;

		// Load height data
		u16* heightmap = new u16[(size+1) * (size+1)];
		fread(heightmap, 2, ((size+1)*(size+1)), f);

		height = new float*[(size+1)];
		for(int i=0; i<(size+1); i++) 
		{
			height[i] = new float[(size+1)];
			for(int j=0; j<(size+1); j++) 
			{
				height[i][j] = (float) (heightmap[(j*(size+1)) + i] / 256.0f) * 0.35f;
			}
		}


		// Load the list of used textures

		int numTextures;
		int strLength;

		fread(&numTextures, sizeof(int), 1, f);

		for (int i=0; i<numTextures; i++)
		{
			fread(&strLength, sizeof(int), 1, f);
			std::string name;

			vector<char> buf(strLength+1);
			fread(&buf[0], 1, strLength, f);
			name = &buf[0];

			// This will add the texture to the NameToId and idToName vectors if it 
			// doesn't already exist. And in this case it shouldn't. 
			getId( name.substr(0, name.length()-4));
		}

		texture = new int*[size];
		for(int i=0; i<size; i++) {
			texture[i] = new int[size];
		}

		Tile* tiles = new Tile[size*size];
		fread(tiles, sizeof(Tile), size*size, f);

		for(int x=0; x<size; x++) 
		{
			for(int y=0; y<size; y++) 
			{
				int patchX = x/16, patchY = y/16;
				int offX = x%16, offY = y%16;
				Tile& t = tiles[ (patchY*size/16 + patchX)*16*16 + (offY*16 + offX) ];
				this->texture[x][y] = t.texture1;
			}
		}
		
		terrainObjects = new vector<Object*>*[size];
		for(int i=0; i<size; i++) {
			terrainObjects[i] = new vector<Object*>[size];
		}

		area = new Area**[size];
		for(int i=0; i<size; i++) {
			area[i] = new Area*[size];
			for(int j=0; j<size; j++) {
				area[i][j] = 0;
			}
		}

		fclose(f);
	}


	if ((loadLevel & (LOAD_INTERACTIVES | LOAD_NONINTERACTIVES)) != LOAD_NOTHING)
	{

		//TODO: Load xml here

		if (loadLevel & LOAD_INTERACTIVES)
		{
			//printf("Loading interactives..\n");
		}


		if (loadLevel & LOAD_NONINTERACTIVES)
		{
			//printf("Loading non-interactives..\n");
		}

	}


}

Map::~Map() {
	for(int i=0; i<size; i++) {
		delete[] texture[i];
	}
	delete[] texture;

	for(int i=0; i<size; i++) {
		delete[] terrainObjects[i];
	}
	delete[] terrainObjects;

	for(int i=0; i<size+1; i++) {
		delete[] height[i];
	}
	delete[] height;

	for(int i=0; i<size; i++) {
		delete[] area[i];
	}
	delete[] area;

	for(int i=0; i<objects.size(); i++) {
		delete objects[i];
	}
}

int Map::getId(string texture) {
	if(nameToId.find(texture) != nameToId.end()) {
		return nameToId[texture];
	}
	else {
		int newId = nameToId.size();
		nameToId[texture] = newId;
		idToName[newId] = texture;
		return newId;
	}
}

string Map::getTexture(int x, int y) {
	if(!validT(x,y)) JS_ReportError(cx, "getTexture: invalid tile position");
	return idToName[texture[x][y]];
}

void Map::setTexture(int x, int y, const string& t) {
	if(!validT(x,y)) JS_ReportError(cx, "setTexture: invalid tile position");
	texture[x][y] = getId(t);
}

float Map::getHeight(int x, int y) {
	if(!validH(x,y)) JS_ReportError(cx, "getHeight: invalid vertex position");
	return height[x][y];
}

void Map::setHeight(int x, int y, float h) {
	if(!validH(x,y)) JS_ReportError(cx, "setHeight: invalid vertex position");
	height[x][y] = h;
}

vector<Object*> Map::getTerrainObjects(int x, int y) {
	if(!validT(x,y)) JS_ReportError(cx, "getTerrainObjects: invalid tile position");
	return terrainObjects[x][y];
}

void Map::setTerrainObjects(int x, int y, vector<Object*> &objects) {
	if(!validT(x,y)) JS_ReportError(cx, "setTerrainObjects: invalid tile position");
	terrainObjects[x][y] = objects;
}

void Map::placeTerrain(int x, int y, Terrain* t) {
	t->place(this, x, y);
}

void Map::addObject(Object* ent) {
	objects.push_back(ent);
}

Area* Map::createArea(AreaPlacer* placer, AreaPainter* painter, Constraint* constr) {
	vector<Point> points;
	if(!placer->place(this, constr, points)) {
		return 0;
	}
	Area* a = new Area(points);
	for(int i=0; i<points.size(); i++) {
		area[points[i].x][points[i].y] = a;
	}
	painter->paint(this, a);
	areas.push_back(a);
	return a;
}

bool Map::createObjectGroup(ObjectGroupPlacer* placer, int player, Constraint* constr) {
	return placer->place(this, player, constr);
}

int Map::createTileClass() {
	tileClasses.push_back(new TileClass(size));
	return tileClasses.size();
}

float Map::getExactHeight(float x, float y) {
	// copied & modified from ScEd

	int xi = min((int) floor(x), size);
	int yi = min((int) floor(y), size);
	float xf = x - xi;
	float yf = y - yi;

	float h00 = height[xi][yi];
	float h01 = height[xi][yi+1];
	float h10 = height[xi+1][yi];
	float h11 = height[xi+1][yi+1];

	return ( 1 - yf ) * ( ( 1 - xf ) * h00 + xf * h10 ) + yf * ( ( 1 - xf ) * h01 + xf * h11 ) ;
}

File: binaries/data/mods/public/maps/random/rmgen/map.js
/**
 * Class for holding map data and providing basic API to change it
 *
 * @param {int} [size] - Size of the map in tiles
 * @param {float} [baseHeight] - Starting height of the map
 */
function Map(size, baseHeight)
{
	// Size must be 0 to 1024, divisible by patches
	this.size = size;

	// Create 2D arrays for textures, object, and areas
	this.texture = [];
	this.terrainObjects = [];
	this.area = [];

	for (let i = 0; i < size; ++i)
	{
		// Texture IDs
		this.texture[i] = new Uint16Array(size);
		// Entities
		this.terrainObjects[i] = [];
		// Area IDs
		this.area[i] = new Uint16Array(size);

		for (let j = 0; j < size; ++j)
			this.terrainObjects[i][j] = [];
	}

	// Create 2D array for heightmap
	let mapSize = size;
	if (!TILE_CENTERED_HEIGHT_MAP)
		++mapSize;

	this.height = [];
	for (let i = 0; i < mapSize; ++i)
	{
		this.height[i] = new Float32Array(mapSize);

		for (let j = 0; j < mapSize; ++j)
			this.height[i][j] = baseHeight;
	}

	// Create name <-> id maps for textures
	this.nameToID = {};
	this.IDToName = [];

	// Array of objects (entitys/actors)
	this.objects = [];
	// Array of integers
	this.tileClasses = [];

	this.areaID = 0;

	// Starting entity ID, arbitrary number to leave some space for player entities
	this.entityCount = 150;
}

Map.prototype.initTerrain = function(baseTerrain)
{
	for (let i = 0; i < this.size; ++i)
		for (let j = 0; j < this.size; ++j)
			baseTerrain.place(i, j);
};

// Return ID of texture (by name)
Map.prototype.getTextureID = function(texture)
{
	if (texture in this.nameToID)
		return this.nameToID[texture];

	// Add new texture
	let id = this.IDToName.length;
	this.nameToID[texture] = id;
	this.IDToName[id] = texture;

	return id;
};

// Return next free entity ID
Map.prototype.getEntityID = function()
{
	return this.entityCount++;
};

// Check bounds on tile map
Map.prototype.validT = function(x, z, distance = 0)
{
	distance += MAP_BORDER_WIDTH;

	if (g_MapSettings.CircularMap)
	{
		let halfSize = Math.floor(this.size / 2);
		return Math.round(Math.euclidDistance2D(x, z, halfSize, halfSize)) < halfSize - distance - 1;
	}
	else
		return x >= distance && z >= distance && x < this.size - distance && z < this.size - distance;
};

// Check bounds on tile map
Map.prototype.inMapBounds = function(x, z)
{
	return x >= 0 && z >= 0 && x < this.size && z < this.size;
};

// Check bounds on height map if TILE_CENTERED_HEIGHT_MAP==true then it's (size, size) otherwise (size + 1 by size + 1)
Map.prototype.validH = function(x, z)
{
	if (x < 0 || z < 0)
		return false;
	if (TILE_CENTERED_HEIGHT_MAP)
		return x < this.size && z < this.size;
	return x <= this.size && z <= this.size;
};

// Check bounds on tile class
Map.prototype.validClass = function(c)
{
	return c >= 0 && c < this.tileClasses.length;
};

Map.prototype.getTexture = function(x, z)
{
	if (!this.validT(x, z))
		throw new Error("getTexture: invalid tile position (" + x + ", " + z + ")");

	return this.IDToName[this.texture[x][z]];
};

Map.prototype.setTexture = function(x, z, texture)
{
	if (!this.validT(x, z))
		throw new Error("setTexture: invalid tile position (" + x + ", " + z + ")");

	this.texture[x][z] = this.getTextureID(texture);
};

Map.prototype.getHeight = function(x, z)
{
	if (!this.validH(x, z))
		throw new Error("getHeight: invalid vertex position (" + x + ", " + z + ")");

	return this.height[x][z];
};

Map.prototype.setHeight = function(x, z, height)
{
	if (!this.validH(x, z))
		throw new Error("setHeight: invalid vertex position (" + x + ", " + z + ")");

	this.height[x][z] = height;
};

Map.prototype.getTerrainObjects = function(x, z)
{
	if (!this.validT(x, z))
		throw new Error("getTerrainObjects: invalid tile position (" + x + ", " + z + ")");

	return this.terrainObjects[x][z];
};

Map.prototype.setTerrainObject = function(x, z, object)
{
	if (!this.validT(x, z))
		throw new Error("setTerrainObject: invalid tile position (" + x + ", " + z + ")");

	this.terrainObjects[x][z] = object;
};

Map.prototype.placeTerrain = function(x, z, terrain)
{
	terrain.place(x, z);
};

Map.prototype.addObject = function(obj)
{
	this.objects.push(obj);
};

Map.prototype.createArea = function(placer, painter, constraint)
{
	// Check for multiple painters
	if (painter instanceof Array)
		painter = new MultiPainter(painter);

	if (constraint === undefined || constraint === null)
		constraint = new NullConstraint();
	else if (constraint instanceof Array)
		// Check for multiple constraints
		constraint = new AndConstraint(constraint);

	let points = placer.place(constraint);
	if (!points)
		return undefined;

	let newID = ++this.areaID;
	let area = new Area(points, newID);
	for (let p of points)
		this.area[p.x][p.z] = newID;

	painter.paint(area);

	return area;
};

Map.prototype.createObjectGroup = function(placer, player, constraint)
{
	// Check for null constraint
	if (constraint === undefined || constraint === null)
		constraint = new NullConstraint();
	else if (constraint instanceof Array)
		constraint = new AndConstraint(constraint);

	return placer.place(player, constraint);
};

Map.prototype.createTileClass = function()
{
	let newID = this.tileClasses.length;
	this.tileClasses.push(new TileClass(this.size, newID));

	return newID;
};

// Get height taking into account terrain curvature
Map.prototype.getExactHeight = function(x, z)
{
	let xi = Math.min(Math.floor(x), this.size);
	let zi = Math.min(Math.floor(z), this.size);
	let xf = x - xi;
	let zf = z - zi;

	let h00 = this.height[xi][zi];
	let h01 = this.height[xi][zi + 1];
	let h10 = this.height[xi + 1][zi];
	let h11 = this.height[xi + 1][zi + 1];

	return (1 - zf) * ((1 - xf) * h00 + xf * h10) + zf * ((1 - xf) * h01 + xf * h11);
};

// Converts from the tile centered height map to the corner based height map, used when TILE_CENTERED_HEIGHT_MAP = true
Map.prototype.cornerHeight = function(x, z)
{
	let count = 0;
	let sumHeight = 0;

	for (let dir of [[-1, -1], [-1, 0], [0, -1], [0, 0]])
		if (this.validH(x + dir[0], z + dir[1]))
		{
			++count;
			sumHeight += this.height[x + dir[0]][z + dir[1]];
		}

	if (count == 0)
		return 0;

	return sumHeight / count;
};

Map.prototype.getFullEntityList = function(rotateForMapExport = false)
{
	// Change rotation from simple 2d to 3d befor giving to engine
	if (rotateForMapExport)
		for (let obj of this.objects)
			obj.rotation.y = PI / 2 - obj.rotation.y;

	// All non terrain objects
	let entities = this.objects;

	// Terrain objects e.g. trees
	let size = this.size;
	for (let x = 0; x < size; ++x)
		for (let z = 0; z < size; ++z)
			if (this.terrainObjects[x][z] !== undefined)
				entities.push(this.terrainObjects[x][z]);

	return entities;
};

Map.prototype.getMapData = function()
{
	let data = {};

	// Convert 2D heightmap array to flat array
	// Flat because it's easier to handle by the engine
	let mapSize = this.size + 1;
	let height = new Uint16Array(mapSize * mapSize);
	for (let x = 0; x < mapSize; ++x)
		for (let z = 0; z < mapSize; ++z)
		{
			let currentHeight;
			if (TILE_CENTERED_HEIGHT_MAP)
				currentHeight = this.cornerHeight(x, z);
			else
				currentHeight = this.height[x][z];

			// Correct height by SEA_LEVEL and prevent under/overflow in terrain data
			height[z * mapSize + x] = Math.max(0, Math.min(0xFFFF, Math.floor((currentHeight + SEA_LEVEL) * HEIGHT_UNITS_PER_METRE)));
		}

	data.height = height;
	data.seaLevel = SEA_LEVEL;

	// Terrain, map width in tiles
	data.size = this.size;

	// Get array of textures used in this map
	data.textureNames = this.IDToName;

	// Entities
	data.entities = this.getFullEntityList(true);
	log("Number of entities: "+ data.entities.length);

	//  Convert 2D tile data to flat array
	let tileIndex = new Uint16Array(this.size * this.size);
	let tilePriority = new Uint16Array(this.size * this.size);
	for (let x = 0; x < this.size; ++x)
		for (let z = 0; z < this.size; ++z)
		{
			// TODO: For now just use the texture's index as priority, might want to do this another way
			tileIndex[z * this.size + x] = this.texture[x][z];
			tilePriority[z * this.size + x] = this.texture[x][z];
		}

	data.tileData = { "index": tileIndex, "priority": tilePriority };

	return data;
};
